lightning/src/ln/channelmanager.rs

```diff
@@ -7389,23 +7538,33 @@ where
 
 	fn peer_disconnected(&self, counterparty_node_id: &PublicKey) {
 		let _persistence_guard = PersistenceNotifierGuard::notify_on_drop(self);
-		let mut failed_channels = Vec::new();
 		let mut per_peer_state = self.per_peer_state.write().unwrap();
 		let remove_peer = {
 			log_debug!(self.logger, "Marking channels with {} disconnected and generating channel_updates.",
 				log_pubkey!(counterparty_node_id));
 			if let Some(peer_state_mutex) = per_peer_state.get(counterparty_node_id) {
 				let mut peer_state_lock = peer_state_mutex.lock().unwrap();
 				let peer_state = &mut *peer_state_lock;
-				let pending_msg_events = &mut peer_state.pending_msg_events;
+
 				peer_state.channel_by_id.retain(|_, chan| {
-					chan.remove_uncommitted_htlcs_and_mark_paused(&self.logger);
-					if chan.is_shutdown() {
+					if !chan.context.is_funding_broadcast() {
 						update_maps_on_chan_removal!(self, &chan.context);
 						self.issue_channel_close_events(&chan.context, ClosureReason::DisconnectedPeer);
-						return false;
+						// It is possible to have persisted the monitor upon funding_signed
+						// but not have broadcast the transaction, especially for batch funding.
+						// The monitor should be moved to the correct state.
+						self.finish_force_close_channel(chan.context.force_shutdown(false));
```

Comment:
The `remove_uncommitted_htlcs_and_mark_paused` and `force_shutdown` seem mutually exclusive though as they both assert that the channel hasn't been shut down yet. In this case, we would like to close the monitor as well.

> Also, if we remove a channel from a batch here, don't we also need to remove the other channels from the same batch (which could be with a different peer not being disconnected)?

Yes, this currently happens in the `update_maps_on_chan_removal` macro so that it will be handled in all places where we close channels.

Reply:
With regard to mutual exclusivity, both functions assert that the function is not in the `ShutdownComplete` state and then move the channel to that state, so I think we can't call both functions?

https://github.com/lightningdevkit/rust-lightning/blob/f6a45056791b2dbbd0b41f581080d9884cda1e3a/lightning/src/ln/channel.rs#L1974
https://github.com/lightningdevkit/rust-lightning/blob/f6a45056791b2dbbd0b41f581080d9884cda1e3a/lightning/src/ln/channel.rs#L2005
https://github.com/lightningdevkit/rust-lightning/blob/f6a45056791b2dbbd0b41f581080d9884cda1e3a/lightning/src/ln/channel.rs#L3589
https://github.com/lightningdevkit/rust-lightning/blob/f6a45056791b2dbbd0b41f581080d9884cda1e3a/lightning/src/ln/channel.rs#L3591